src/bin/pg_autoctl/monitor.c

```diff
@@ -1190,3 +1199,261 @@ monitor_get_notifications(Monitor *monitor)
+				  PG_AUTOCTL_MONITOR_EXTENSION_NAME);
+		return false;
+	}
+
+	if (strcmp(version.installedVersion, PG_AUTOCTL_EXTENSION_VERSION) != 0)
+	{
+		Monitor dbOwnerMonitor = { 0 };
+
+		log_warn("This version of pg_autoctl requires the extension \"%s\" "
+				 "version \"%s\" to be installed on the monitor, current "
+				 "version is \"%s\".",
+				 PG_AUTOCTL_MONITOR_EXTENSION_NAME,
+				 PG_AUTOCTL_EXTENSION_VERSION,
+				 version.installedVersion);
+
+		/*
+		 * Ok, let's try to update the extension then.
+		 *
+		 * For that we need to connect as the owner of the database, which was
+		 * the current $USER at the time of the `pg_autoctl create monitor`
+		 * command.
+		 */
+		if (!prepare_connection_to_current_system_user(monitor,
+													   &dbOwnerMonitor))
+		{
+			log_error("Failed to upgrade extension \"%s\" to version \"%s\": "
+					  "failed prepare a connection string to the "
+					  "monitor as the database owner",
+					  PG_AUTOCTL_MONITOR_EXTENSION_NAME,
+					  PG_AUTOCTL_EXTENSION_VERSION);
+			return false;
+		}
+
+		if (!monitor_extension_update(&dbOwnerMonitor,
+									  PG_AUTOCTL_EXTENSION_VERSION))
+		{
+			log_fatal("Failed to update extension \"%s\" to version \"%s\" "
+					  "on the monitor, see above for details",
+					  PG_AUTOCTL_MONITOR_EXTENSION_NAME,
+					  PG_AUTOCTL_EXTENSION_VERSION);
+			return false;
+		}
+
+		log_info("Updated extension \"%s\" to version \"%s\"",
+				 PG_AUTOCTL_MONITOR_EXTENSION_NAME,
+				 PG_AUTOCTL_EXTENSION_VERSION);
+
+		return true;
+	}
+
+	/* just mention we checked, and it's ok */
+	log_info("The version of extenstion \"%s\" is \"%s\" on the monitor",
+			 PG_AUTOCTL_MONITOR_EXTENSION_NAME, version.installedVersion);
+
+	return true;
+}
+
+
+/*
+ * prepare_connection_to_current_system_user changes a given pguri to remove
+ * its "user" connection parameter, filling in the pre-allocated keywords and
+ * values string arrays.
+ *
+ * Postgres docs at the following address show 30 connection parameters, so the
+ * arrays should allocate 31 entries at least. The last one is going to be
+ * NULL.
+ *
+ *  https://www.postgresql.org/docs/current/libpq-connect.html
+ */
+static bool
+prepare_connection_to_current_system_user(Monitor *source, Monitor *target)
+{
+	const char *keywords[31] = { 0 };
+	const char *values[31] = { 0 };
+
+	char *errmsg;
+	PQconninfoOption *conninfo, *option;
+	int argCount = 0;
+
+	conninfo = PQconninfoParse(source->pgsql.connectionString, &errmsg);
+	if (conninfo == NULL)
+	{
+		log_error("Failed to parse pguri \"%s\": %s",
+				  source->pgsql.connectionString, errmsg);
+		PQfreemem(errmsg);
+		return false;
+	}
+
+	for (option = conninfo; option->keyword != NULL; option++)
+	{
+		if (strcmp(option->keyword, "user") == 0)
+		{
+			/* skip the user, $USER is what we want to use here */
+			continue;
+		}
+		else if (option->val)
+		{
+			keywords[argCount] = option->keyword;
+			values[argCount] = option->val;
+			++argCount;
```

Comment:
lets be defensive here and error when we go over the size of the array.

Reply:
Fixed.